prov/lnx/src/lnx_init.c

```diff
@@ -0,0 +1,881 @@
+	.size = sizeof(struct fi_ops_fabric),
+	.domain = lnx_domain_open,
+	.passive_ep = fi_no_passive_ep,
+	.eq_open = fi_no_eq_open,
+	.wait_open = fi_no_wait_open,
+	.trywait = fi_no_trywait
+};
+
+struct fi_provider lnx_prov = {
+	.name = OFI_LNX,
+	.version = OFI_VERSION_DEF_PROV,
+	.fi_version = OFI_VERSION_LATEST,
+	.getinfo = lnx_getinfo,
+	.fabric = lnx_fabric,
+	.cleanup = lnx_fini
+};
+
+struct util_prov lnx_util_prov = {
+	.prov = &lnx_prov,
+	.info = &lnx_info,
+	.flags = 0
+};
+
+/*
+ * For the fi_getinfo() -> fi_fabric() -> fi_domain() path, we need to
+ * keep track of the fi_info in case we need them later on when linking in
+ * the fi_fabric() function.
+ *
+ * This cache gets cleared after we use the ones we need, or when the
+ * library exists, if LNX is never used.
+ */
+struct dlist_entry lnx_fi_info_cache;
+/* this is a list of all possible links */
+struct dlist_entry lnx_links;
+struct dlist_entry lnx_links_meta;
+
+struct lnx_fi_cache_entry {
+	struct dlist_entry entry;
+	struct fi_info *fi;
+};
+
+struct lnx_fi_info_meta {
+	struct dlist_entry entry;
+	struct fi_info *lnx_rep;
+	struct fi_info *lnx_link;
+};
+
+static int lnx_get_cache_meta(struct dlist_entry *head, int *size)
+{
+	int num_prov = 0;
+	struct dlist_entry *e;
+
+	dlist_foreach(head, e)
+		num_prov++;
+
+	*size = num_prov;
+
+	return FI_SUCCESS;
+}
+
+static void lnx_free_meta(void)
+{
+	struct lnx_fi_info_meta *e;
+	struct dlist_entry *tmp;
+
+	dlist_foreach_container_safe(&lnx_links_meta, struct lnx_fi_info_meta, e,
+				     entry, tmp) {
+		dlist_remove(&e->entry);
+		free(e);
+	}
+}
+
+static void lnx_free_info_cache(struct dlist_entry *head, bool meta)
+{
+	struct lnx_fi_cache_entry *e;
+	struct dlist_entry *tmp;
+
+	dlist_foreach_container_safe(head, struct lnx_fi_cache_entry, e,
+				     entry, tmp) {
+		fi_freeinfo(e->fi);
+		dlist_remove(&e->entry);
+		free(e);
+	}
+
+	if (meta)
+		lnx_free_meta();
+}
+
+static int lnx_cache_info(struct dlist_entry *head,
+			  struct fi_info *info)
+{
+	struct lnx_fi_cache_entry *e = calloc(1, sizeof(*e));
+	if (!e)
+		return -FI_ENOMEM;
+	dlist_init(&e->entry);
+	e->fi = info;
+
+	dlist_insert_tail(&e->entry, head);
+
+	return 0;
+}
+
+struct fi_info *
+lnx_get_link_by_dom(char *domain_name)
+{
+	struct fi_info *info;
+	struct lnx_fi_info_meta *e;
+
+	dlist_foreach_container(&lnx_links_meta, struct lnx_fi_info_meta, e,
+				entry) {
+		info = e->lnx_rep;
+		if (info && info->domain_attr) {
+			if (!strcmp(domain_name,
+				    info->domain_attr->name)) {
+				FI_INFO(&lnx_prov, FI_LOG_CORE, "Found %s\n",
+						info->fabric_attr->prov_name);
+				return e->lnx_link;
+			}
+		}
+	}
+
+	return NULL;
+}
+
+static void lnx_insert_tail(struct fi_info *head, struct fi_info *item)
+{
+	struct fi_info *itr = head;
+	while(itr->next)
+		itr = itr->next;
+	itr->next = item;
+}
+
+static void lnx_remove_tail(struct fi_info **head)
+{
+	struct fi_info *itr = *head, *prev = NULL;
+	while(itr->next) {
```

Comment:
*** space after while, new line between declares and while

Reply:
done.